src/client/pythonEnvironments/discovery/locators/services/condaLocator.ts

```diff
@@ -1,125 +1,72 @@
+    // Locating conda binary is expensive, since it potentially involves spawning or
+    // trying to spawn processes; so it's done lazily and asynchronously. Methods that
+    // need a Conda instance should use getConda() to obtain it, and should never access
+    // this property directly.
+    private condaPromise: Promise<Conda | undefined> | undefined;
 
-    // Check if the conda-meta directory is in the same directory as the interpreter.
-    // This layout is common in Windows.
-    // env
-    // |__ conda-meta  <--- check if this directory exists
-    // |__ python.exe  <--- interpreterPath
-    const condaEnvDir1 = path.join(path.dirname(interpreterPath), condaMetaDir);
-
-    // Check if the conda-meta directory is in the parent directory relative to the interpreter.
-    // This layout is common on linux/Mac.
-    // env
-    // |__ conda-meta  <--- check if this directory exists
-    // |__ bin
-    //     |__ python  <--- interpreterPath
-    const condaEnvDir2 = path.join(path.dirname(path.dirname(interpreterPath)), condaMetaDir);
-
-    // The paths are ordered in the most common to least common
-    return [condaEnvDir1, condaEnvDir2];
-}
-
-/**
- * Checks if the given interpreter path belongs to a conda environment. Using
- * known folder layout, and presence of 'conda-meta' directory.
- * @param {string} interpreterPath: Absolute path to any python interpreter.
- *
- * Remarks: This is what we will use to begin with. Another approach we can take
- * here is to parse ~/.conda/environments.txt. This file will have list of conda
- * environments. We can compare the interpreter path against the paths in that file.
- * We don't want to rely on this file because it is an implementation detail of
- * conda. If it turns out that the layout based identification is not sufficient
- * that is the next alternative that is cheap.
- *
- * sample content of the ~/.conda/environments.txt:
- * C:\envs\myenv
- * C:\ProgramData\Miniconda3
- *
- * Yet another approach is to use `conda env list --json` and compare the returned env
- * list to see if the given interpreter path belongs to any of the returned environments.
- * This approach is heavy, and involves running a binary. For now we decided not to
- * take this approach, since it does not look like we need it.
- *
- * sample output from `conda env list --json`:
- * conda env list --json
- * {
- *   "envs": [
- *     "C:\\envs\\myenv",
- *     "C:\\ProgramData\\Miniconda3"
- *   ]
- * }
- */
-export async function isCondaEnvironment(interpreterPath: string): Promise<boolean> {
-    const condaMetaPaths = getCondaMetaPaths(interpreterPath);
-    // We don't need to test all at once, testing each one here
-    for (const condaMeta of condaMetaPaths) {
-        if (await pathExists(condaMeta)) {
-            return true;
+    public constructor(conda?: Conda) {
+        super();
+        if (conda !== undefined) {
+            this.condaPromise = Promise.resolve(conda);
```

Comment:
Would it make sense to discover it in the background here, to get the ball rolling ASAP?

```suggestion
        if (conda !== undefined) {
            this.condaPromise = Promise.resolve(conda);
        } else {
            this.condaPromise = Conda.locate();
```

Note that doing so makes `this.getConda()` unnecessary and means `condaPromise` can be just `Promise<Conda | undefined>`.

Reply:
That's what I did originally, but it didn't quite work out - `Conda.locate()` needs to be able to spawn processes, and the implementation of that in `externalDependencies` currently requires the exec service, which is not initialized yet at the time the constructor is invoked. This way, `locate()` only gets called after all services are created.

Reply:
Hmm, that's a bummer.  We are going to have to excise the IOC "services" at some point, so now may be a suitable time to update the `externalDependencies` implementation to use node's `childProcess` module directly.  Of course, if doing so would be much work then we could queue up a separate task to do so (and to do the early `Conda.locate()` call).